File: backend/src/main/java/com/seoulawake/backend/module/cafe/application/CafeService.java
package com.seoulawake.backend.module.cafe.application;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.seoulawake.backend.infra.kakao.KakaoApiTemplate;
import com.seoulawake.backend.infra.kakao.dto.request.LocalGetAddressRequest;
import com.seoulawake.backend.infra.kakao.dto.response.LocalGetAddressResponse;
import com.seoulawake.backend.module.cafe.domain.Cafe;
import com.seoulawake.backend.module.cafe.domain.repository.CafeRepository;

import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;

@Service
@Transactional
@RequiredArgsConstructor
public class CafeService {
	private final CafeRepository cafeRepository;
	private final KakaoApiTemplate kakaoApiTemplate;

	@Transactional(readOnly = true)
	public List<Cafe> getCafes() {
		return cafeRepository.findAll();
	}

	public Cafe create(String address) {
		LocalGetAddressRequest request = new LocalGetAddressRequest(address);
		LocalGetAddressResponse response = kakaoApiTemplate.execute(request, LocalGetAddressResponse.class);

		Cafe cafe = new Cafe(null, null); // todo: apply api result
		return cafeRepository.save(cafe);
	}

	public boolean checked(Long id) {
		Cafe cafe = findById(id);
		cafe.checked();
		return true;
	}

	public Cafe getCafe(String name) {
		return cafeRepository.findByName(name)
			.orElseThrow(EntityNotFoundException::new);
	}

	private Cafe findById(Long id) {
		return cafeRepository.findById(id)
			.orElseThrow(EntityNotFoundException::new);
	}
}

File: backend/src/main/kotlin/seoulawake/server/module/cafe/application/CafeService.kt
package seoulawake.server.module.cafe.application

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import seoulawake.server.global.exception.ErrorCode.ALREADY_REGISTERED_CAFE
import seoulawake.server.global.exception.validate
import seoulawake.server.module.cafe.domain.Address
import seoulawake.server.module.cafe.domain.Cafe
import seoulawake.server.module.cafe.domain.CafeRepository
import seoulawake.server.module.cafe.domain.Coordinates
import seoulawake.server.module.cafe.dto.CafeDetails
import seoulawake.server.module.cafe.dto.RegisterCafe

@Service
@Transactional(readOnly = true)
class CafeService(
  private val cafeRepository: CafeRepository,
) {

  @Transactional
  fun register(request: RegisterCafe) {
    val (name, address, roadAddress, latitude, longitude) = request
    validateNotRegistered(name, address)
    val cafe = Cafe(name, address = Address(address, roadAddress), coordinates = Coordinates(latitude, longitude))
    cafeRepository.save(cafe)
  }

  private fun validateNotRegistered(name: String, address: String) {
    validate(!cafeRepository.existsByNameOrAddress_Address(name, address)) { ALREADY_REGISTERED_CAFE }
  }

  fun loadAllCafes(): List<CafeDetails> {
    return cafeRepository.findAll().map {
      CafeDetails(
        it.name,
        it.address.address,
        it.address.roadAddress,
        it.coordinates.latitude,
        it.coordinates.longitude,
        it.status.verified
      )
    }.toList()
  }
}
